hw1

tiginamaria opened the pull request.

tiginamaria committed:
hw1

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     * Constructor for HashTable with default capacity = 1
+     */
+    public HashTable() {
+        size = 0;
```
Вообще, оно и так проинициализируется нулём при создании объекта, так что это можно не писать. Но можно и писать

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     * Constructor for HashTable with given capacity
+     * @param cp - capacity
+     */
+    public HashTable(int cp) {
```
Так и надо было его назвать capacity. Имена параметров --- это святое, потому что они показываются во всплывающих подсказках. Ниже можно было this.capacity = capacity;

Вообще, лучше никогда не сокращать идентификаторы.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+        table = new List[capacity];
+        for (int i = 0; i < capacity; i++) {
+            table[i] = new List();
+        }
```
Что-то копипаст. В Java тоже есть constructor chaining.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+    /**
+     * Double the capacity of HashTable
+     */
+    private void extand() {
```
ex**p**and? или ext**e**nd? :)

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+        return size;
+    }
+
+    public int getCapacity() {
```
А зачем бы внешнему миру знать capacity, если хеш-таблица всё равно ей сама управляет?

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    private void extand() {
+        int newCapacity = capacity * 2;
+        List[] newTable = new List[newCapacity];
```
Тут лучше var. Не нужно писать тип дважды. Вообще, если тип переменной очевиден из правой части, правильнее писать var.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+            newTable[i] = new List();
+        }
+
+        for(int i = 0; i != capacity; i++) {
```
Надо пробел между for и скобкой

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+        }
+
+        for(int i = 0; i != capacity; i++) {
+            while(!table[i].empty()) {
```
И тут, и в некоторых местах ниже

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     * @return hash for given key
+     */
+    private int getHash(String key, int mod) {
+        return key.hashCode() % mod;
```
hashCode может возвращать отрицательные значения, % тоже, а код ниже к этому морально не готов. Так что оно иногда будет падать --- на длинных строках, некоторых строках национальных алфавитов и т.д.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public String get(String key) {
+        if(key == null)
+            return null;
```
Надо фигурные скобки вокруг тел операторов, даже если там один оператор, во избежание глупых ошибок, тут и ниже

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     * @return Previous value of the element with given key or null if it does not exist
+     */
+    public String put(String key, String value) {
+        String elem = table[getHash(key, capacity)].put(key, value);
```
Если в качестве key передадут null, будет NullReferenceException, который указывает на то, что программист о чём-то не подумал. А я бы ожидал IllegalArgumentException, который указывает на то, что о чём-то не подумал пользователь

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+    public String put(String key, String value) {
+        String elem = table[getHash(key, capacity)].put(key, value);
+        if(elem == null)
+            size++;
```
Хм, а если там правда был элемент со значением null? put не запрещает null в value

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+     * @return Previous value of the element with given key or null if it does not exist
+     */
+    public String remove(String key) {
+        String elem = table[getHash(key, capacity)].remove(key);
```
Тоже не очень дружит с null-ом

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,133 @@
+        for(int i = 0; i < capacity; i++) {
+            table[i].clear();
+        }
+        size = 0;
```
Изменение размера таблицы может привести к тому, что массив, хранящий списки, станет очень большим. clear тут с этим ничего не делает, поэтому может быть неожиданная для пользователя ситуация, которая очень похожа на утечку памяти: мы положили в таблицу миллион элементов, увидели, что у нас скушалась вся оперативка, сделали clear и обещали себе никогда больше 10 элементов в таблицу не класть. А память-то всё равно расходуется так, будто в таблице миллион элементов.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/List.java:
```diff
@@ -0,0 +1,145 @@
+         * Constructor for Node(element of list)
+         * @param d - data
+         */
+        public Node(Data d) {
```
Тут тоже, имена параметров лучше сделать покрасивее.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/List.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Node is an element of list, which contains information(data) and links to next and previous elements of list
+     */
+    private class Node {
```
Вообще, он должен быть static, но вам пока про это не рассказывали, так что пока можно не править. Подробности: https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/List.java:
```diff
@@ -0,0 +1,145 @@
+
+    /**
+     * Check if list is empty
+     * @return true - if empty, false - if not empty
```
Совсем уж капитанские комментарии можно не писать

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/List.java:
```diff
@@ -0,0 +1,145 @@
+     * @return Node  with given key or null if the key has not been found
+     */
+    public Node find(String key) {
+        Node cur = head;
```
Тоже, лучше не сокращать имена, хотя тут область видимости переменной очень узка, так что это ещё хоть как-то оправданно. Это Java, а не олимпиадные C :)

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/HashTableTest.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Test
+    void putGetRemoveHashTableTest() {
+        HashTable t = new HashTable();
```
Имеет смысл использовать @BeforeEach для инициализации хеш-таблицы, а то копипаст

tiginamaria committed:
fixed all codestyle mistakes

tiginamaria committed:
fixed null key and value mistakes
added exceptions for null key and value
added new tests for exceptions
added preinitialisation (@BeforeEach) to file with tests
fixed .clear() for HashTable

tiginamaria committed:
added exceptions for null key and value in .getHashcode() in HashTable.java

tiginamaria committed:
fixed small slips in names and annotations
fixed mistake in .removeFromHead() in class List

tiginamaria committed:
new .idea files

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,152 @@
+
+    /**
+     * Calculate hash for given key
+     * @param key - given key of element
```
На самом деле, "-" лучше не писать, потому что его генерит JavaDoc и так получится, что он будет в документации дважды.

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,152 @@
+            throw new IllegalArgumentException("key can not be null!");
+        }
+        int hash = key.hashCode() % mod;
+        if(hash < 0) {
```
Пробела не хватает

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/HashTable.java:
```diff
@@ -0,0 +1,152 @@
+
+        String oldData = table[getHash(key, capacity)].remove(key);
+        if (oldData != null)
+            size--;
```
Тут надо ещё фигурные скобки

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/List.java:
```diff
@@ -0,0 +1,146 @@
+     */
+    public Data removeFromHead() {
+        if (head == null)
+            return null;
```
И тут фигурные скобки лучше писать

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/DataTest.java:
```diff
@@ -0,0 +1,38 @@
+
+class DataTest {
+
+    private Data d;
```
Не лучшее название для поля

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/DataTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @BeforeEach
+    void initData() {
+        d = new Data("a", "b");;
```
Лишняя `;` :)

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/DataTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    void getValueDataTest() {
+        assertEquals(d.getValue(),"b");
```
А тут пробела после запятой не хватает

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/HashTableTest.java:
```diff
@@ -0,0 +1,102 @@
+
+class HashTableTest {
+
+    private HashTable t;
```
Тоже, имя поля не очень. Общая рекомендация --- избегать имён полей короче 4 символов.

tiginamaria committed:
fixed code style mistakes
changed bad names for class's fields

yurii-litvinov commented on hashtable/src/main/java/ru/hse/java/hashtable/Data.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Constructor for Data
+     * @param key key
+     * @param value value
```
Очевидные комментарии можно вообще не писать, они только место занимают.

yurii-litvinov commented on hashtable/src/test/java/ru/hse/java/hashtable/HashTableTest.java:
```diff
@@ -0,0 +1,102 @@
+    void putGetHashTableTest() {
+        hashtable.put("a", "b");
+        hashtable.put("a", "c");
+        assertEquals(hashtable.get("a"), "c");
```
Наоборот, сначала ожидаемое значение, затем то, что получилось. Это влияет на вывод в логе тестирования.

yurii-litvinov reviewed (approved):
Окей, зачтена.

tiginamaria committed:
rename file
